Write metrics to result file during eval

abaek opened the pull request:
If a user specifies a `result` parameter in the eval part of their config, then it will write all metrics to that file.

To: @jq @yinlou

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -139,6 +139,23 @@ object GenericPipeline {
     metrics.foreach(x => log.info(if (x._1 contains "THRESHOLD") x._1 else x.toString))
+
+    // Write metrics to result
+    try {
```
please wrap it into class.

yinlou replied:
agreed. Could be something like this:

def outputResults(records to write): Unit = {
  // log
  records.map(e => log.info(e.toString))
  // Write to hdfs
  Your logic
}

abaek replied:
If I wrap it into a class, im not sure if i should also move the logging there

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -139,6 +139,23 @@ object GenericPipeline {
+
+    // Write metrics to result
+    try {
+      val result: String = confi.getConfig(cfgKey).getString("result")
```
you will need the config of output result path, i.e. if it has output_result_path, then save result this way

abaek commented:
Build doesn't seem to pass, but could you verify that this is the right place to put the code. I can fix the syntax afterwards. @jq

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -156,6 +156,7 @@ object GenericPipeline {
     val isRegression = Try(cfg.getBoolean("is_regression")).getOrElse(false)
     val isMulticlass = Try(cfg.getBoolean("is_multiclass")).getOrElse(false)
     val metric = cfg.getString("metric_to_maximize")
+    val resultsOutputPath = Try(cfg.getString("results_output_path"))..getOrElse("")
```
..

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -447,7 +449,8 @@ object GenericPipeline {
       metric: String,
-      isTraining: Example => Boolean) : Array[(String, Double)] = {
+      isTraining: Example,
+      resultsOutputPath: String => Boolean) : Array[(String, Double)] = {
```
Example => Boolean and String

abaek replied:
I'm not sure what you mean, I thought this line was correct

abaek replied:
nvm, i  understand

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala:
```diff
@@ -87,6 +87,24 @@ object Evaluation {
+  def evaluateBinaryClassification(records : List[EvaluationRecord],
+                                   buckets : Int,
+                                   evalMetric : String,
+                                   resultOutputPath: String) : Array[(String, Double)] = {
```
this two are same?

abaek replied:
One of them is for List and the other is for RDD

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala:
```diff
@@ -0,0 +1,25 @@
+ */
+object ResultUtil {
+  def writeResults(metrics: Array[(String, Double)], resultOutputPath: String) = {
+    val fileSystem = FileSystem.get(new java.net.URI(result), new Configuration())
```
try reuse https://github.com/airbnb/aerosolve/blob/master/training/src/main/scala/com/airbnb/aerosolve/training/pipeline/PipelineUtil.scala#L39

abaek replied:
`PipelineUtil.saveAndCommitAsTextFile` didn't have exactly what i needed, but I used `PipelineUtil.writeStringToFile`

jq commented:
ok to put it in current places. You may talk to yizheng about the format of result.

abaek commented:
Failed a test when I ran `gradle test`, but I don't think it was related to my changes

<img width="854" alt="screen shot 2016-07-26 at 3 50 16 pm" src="https://cloud.githubusercontent.com/assets/5884612/17158037/5b3b94ba-5348-11e6-8f06-2d30921a7241.png">

abaek commented:
Actually, working on changing the format, don't review yet please

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -156,6 +156,7 @@ object GenericPipeline {
     val isRegression = Try(cfg.getBoolean("is_regression")).getOrElse(false)
     val isMulticlass = Try(cfg.getBoolean("is_multiclass")).getOrElse(false)
     val metric = cfg.getString("metric_to_maximize")
+    val resultsOutputPath = Try(cfg.getString("results_output_path")).getOrElse("")
```
please update the demo with results_output_path, this is a good test for the code too.

abaek replied:
done

jq replied:
seems not yet push? i.e. demo https://github.com/airbnb/aerosolve/tree/master/demo/income_prediction

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala:
```diff
@@ -0,0 +1,19 @@
+  def writeResults(metrics: Array[(String, Double)], resultOutputPath: String) = {
+    // Put into a one line json format
+    val elements = metrics.map{ x =>
+      val metricStr = if (x._1 contains "THRESHOLD") x._1 else x.toString
```
move THRESHOLD out, and add Unit test and comments for the output format.

abaek replied:
I added comments for the output format, but I wasn't sure what to test in a unit test

abaek commented:
I changed the result output format and updated this PR. Apparently you can't have default arguments and overload method names in scala, so I created new methods and resued a lot of code from the old methods.

I still have to work on adding a unit test, and testing this code with the demo.

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala:
```diff
@@ -87,6 +88,95 @@ object Evaluation {
+                                   buckets : Int,
+                                   evalMetric : String,
+                                   resultsOutputPath: String) : Array[(String, Double)] = {
+    var metrics  = mutable.Buffer[(String, Double)]()
```
lots of code are dup with the other evaluateBinaryClassification, can you share code between them?

abaek replied:
done, code is never repeated

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala:
```diff
@@ -197,6 +330,16 @@ object Evaluation {
     metrics.append(("!TRAIN_PRECISION", if (ttp > 0.0) ttp / (ttp + tfp) else 0.0))
     metrics.append(("!TRAIN_F1", 2 * ttp / (2 * ttp + tfn + tfp)))
 
+    if (results != null) {
```
maybe it is better to report it with metrics.append, i.e. metricsAppend(metrics, results, name, value)

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/GenericPipeline.scala:
```diff
@@ -470,7 +473,7 @@ object GenericPipeline {
       Evaluation.evaluateMulticlassClassification(records)
     } else {
-      Evaluation.evaluateBinaryClassification(records, bins, metric)
+      Evaluation.evaluateBinaryClassificationWithResults(records, bins, metric, resultsOutputPath)
```
if resultsOutputPath is null, appendMetricAndResult may crash..

abaek replied:
I got rid of `appendMetricAndResult`. Now I am just relying on the metrics buffer and `resultsOutputPath` is only being used in one place here `Evaluation#L78`: 
```
    if (resultsOutputPath != null) {
      ResultUtil.writeResults(resultsOutputPath, metrics.toArray, holdThresholdPrecisionRecall.toArray, trainThresholdPrecisionRecall.toArray)
    }
```

jq replied:
nice, please push the update

abaek replied:
Woops done

jq commented on training/src/main/scala/com/airbnb/aerosolve/training/pipeline/ResultUtil.scala:
```diff
@@ -0,0 +1,113 @@
+ *
+ */
+object ResultUtil {
+  def writeResults(resultsOutputPath: String,
```
suggest you add unit test for ResultUtil in this PR, and make ResultUtil extendable on next PR

jq commented:
👍

saurfang commented on training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala:
```diff
@@ -20,119 +20,25 @@ object Evaluation {
-      .toArray
+    // Convert RDD to list
+    val recordsList = List.fromArray(records.collect());
+    evaluateBinaryClassificationWithResults(recordsList, buckets, evalMetric, resultsOutputPath)
```
@abaek what was the motivation to call `collect` here? was it just to unify the codepath? I thought there were two evaluations that can compute locally when result is small but distributed via RDD when dataset is large.

abaek replied:
The motivation was to make all of the `evaluateBinaryClassification` share one code path and eliminate duplicate code. When I tried to refactor the code, I had a hard time sharing code between RDD's and List.

@jq mentioned that Evaluation jobs were small enough that collecting the RDD was ok.

saurfang replied:
I see. It makes sense but I'm having trouble collecting them to driver and driver is struggling with the evaluation. I'm going to try making everything follow the RDD codepath instead.

abaek replied:
Oh ok sorry about that! I think I only tested it with Lists.
